wire checkEmbedderMeta guard into RecallWithEvent path (split from #756)

## Problem

The dim-mismatch guard in `internal/embed` (`checkEmbedderMeta`) is not currently invoked during `RecallWithEvent`. As a result, when an embedder dimension changes between store and recall, `RecallWithEvent` returns `nil` rather than a typed error.

This was uncovered by PR #756's `TestRetrievalTracking_DimMismatch`, which asserts that `RecallWithEvent` returns an error on dim mismatch. The test fails in CI:

```
--- FAIL: TestRetrievalTracking_DimMismatch (0.05s)
    retrieval_precision_test.go:199:
        Error: An error is expected but got nil.
        Test:  TestRetrievalTracking_DimMismatch
        Messages: RecallWithEvent with dim mismatch must return an error
```

## Root cause

PR #756 was scoped as test-only — asserting an existing guard. The Round 2 (Coverage) reviewer flagged that the guard isn't wired into the recall path. CI confirms it. Without DB-layer rejection (pgvector cross-dim errors), the test cannot pass.

## Fix

Wire `checkEmbedderMeta` (or equivalent dim-check) into `RecallWithEvent` at the entry point, before the embed query is issued. Return a typed error (`ErrEmbedderDimMismatch` or similar) when project metadata indicates an incompatible dim.

## Acceptance

- `RecallWithEvent` returns a typed error on dim mismatch
- `TestRetrievalTracking_DimMismatch` in #756 passes against this wiring
- Unblocks PR #756 for merge

## References

- PR #756 (deferred pending this fix)
- PR #756 Round 2 reviewer comment on coverage path

--- a/internal/search/engine.go
+++ b/internal/search/engine.go
@@ -538,6 +538,14 @@ func (e *SearchEngine) RecallWithOpts(ctx context.Context, query string, topK in
 		topK = 10
 	}
 
+	// Guard: ensure the current embedder matches the stored metadata before
+	// issuing any query. Without this check a dimension mismatch would produce
+	// silent nil/empty results (or a pgvector cross-dim panic) instead of the
+	// structured embed.PermanentError callers can inspect and act on.
+	if err := e.checkEmbedderMeta(ctx); err != nil {
+		return nil, err
+	}
+
 	// Bound the embed call to embedRecallTimeout (default 500ms; configurable via
 	// ENGRAM_EMBED_RECALL_TIMEOUT_MS). On timeout or error, degrade gracefully to
 	// BM25+recency only — the vector leg is skipped but recall still returns useful

--- a/internal/search/retrieval_precision_test.go
+++ b/internal/search/retrieval_precision_test.go
@@ -6,13 +6,29 @@ package search_test
 
 import (
 	"context"
+	"errors"
 	"testing"
 
+	"github.com/petersimmons1972/engram/internal/db"
+	"github.com/petersimmons1972/engram/internal/embed"
 	"github.com/petersimmons1972/engram/internal/search"
 	"github.com/petersimmons1972/engram/internal/types"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
+
+// newEngineWithDims creates a test engine backed by a real Postgres backend
+// using the given embedding dimension. The backend is registered for cleanup.
+// Skips the test if TEST_DATABASE_URL is not set (via testDSN).
+func newEngineWithDims(t *testing.T, proj string, dims int) *search.SearchEngine {
+	t.Helper()
+	ctx := context.Background()
+	backend, err := db.NewPostgresBackend(ctx, proj, testDSN(t))
+	require.NoError(t, err)
+	t.Cleanup(func() { backend.Close() })
+	return search.New(ctx, backend, &fakeClient{dims: dims}, proj,
+		"http://ollama:11434", "llama3.2", false, nil, 0)
+}
 
 // ── Scoring unit tests ────────────────────────────────────────────────────────
 
@@ -147,4 +163,45 @@ func TestRetrievalTracking_PrecisionUpdated(t *testing.T) {
 		"retrieval_precision must be computed once times_retrieved >= 5")
 	assert.InDelta(t, 1.0, *after.RetrievalPrecision, 0.05,
 		"always-useful memory must have precision near 1.0")
+}
+
+// TestRecallWithEvent_ReturnsErrorOnDimMismatch verifies that RecallWithEvent
+// returns a structured embed.PermanentError when the current embedder dimension
+// differs from the dimension stored in project metadata.
+//
+// Regression guard for #767: the store path had checkEmbedderMeta but the
+// recall path did not, so dim mismatches at recall time produced silent nil
+// results instead of the typed error callers need to surface actionable
+// remediation advice.
+func TestRecallWithEvent_ReturnsErrorOnDimMismatch(t *testing.T) {
+	ctx := context.Background()
+	proj := uniqueProject("test-recall-dim-mismatch")
+
+	// Phase 1: store a memory with a 1024-dim engine to seed project metadata.
+	engine1024 := newEngineWithDims(t, proj, 1024)
+	t.Cleanup(func() { engine1024.Close() })
+	m := &types.Memory{
+		Content:     "dim mismatch recall test memory",
+		MemoryType:  types.MemoryTypeDecision,
+		Project:     proj,
+		Importance:  1,
+		StorageMode: "focused",
+	}
+	require.NoError(t, engine1024.Store(ctx, m))
+
+	// Phase 2: attempt recall with a 768-dim engine — must return a clear error,
+	// not silent empty results that would corrupt retrieval precision metrics.
+	engine768 := newEngineWithDims(t, proj, 768)
+	t.Cleanup(func() { engine768.Close() })
+	_, _, err := engine768.RecallWithEvent(ctx, "dim mismatch recall test memory", 5, "normal")
+	require.Error(t, err, "RecallWithEvent with dim mismatch must return an error")
+
+	// Assert on the structured embed.PermanentError — Stored and Current must
+	// reflect the dimensional mismatch (not just an opaque string error).
+	var permErr *embed.PermanentError
+	require.True(t, errors.As(err, &permErr),
+		"error must be an *embed.PermanentError (got: %v)", err)
+	assert.NotEqual(t, permErr.Stored, permErr.Current,
+		"PermanentError must record the dim mismatch in Stored vs Current (stored=%q current=%q)",
+		permErr.Stored, permErr.Current)
 }

